### PR TITLE
76 ip loopback hardware test

### LINKED ISSUE #76: Implement IP loopback
**Feature description**
We would like to demonstrate the IP stack.

**Expected behaviour**
IP packets are echoed back to the sender.

**Expected steps**
1. Update the `rxStack` so that it depacketizes ip packets. Ensure to use the packet dispatcher for the correct eth type and that it filters received IP packets on our IP or any multi/broadcast addresses. 
2. Update the `txStack` so that it packetizes ip packets.
3. Set the destination IP to the source IP and the source IP to our IP. Our IP can be hardcoded.
4. Update the `mac_frame.py` to send an IP packet to the IP of the board and wait for a response.

### PR DESCRIPTION
**Pull request by Akribes**
Expands the echo stack to depacketize and repacketize IP packets and adds a python test to test this. The destination mac address is still hard coded.

Also includes some bug fixes for the up converter and ip depacketizer to make this work.

### CODE REVIEW

**Review by JLaumen — COMMENTED** [review]
We want to provide an example stack (with usage documentation) with just MAC (which currently is Rx/TxStack.hs), so I think it's better to leave those files as is and create new modules that import and build upon those stacks.

File: src/Clash/Cores/Ethernet/RxStack.hs [hunk 1]
```diff
@@ -35,13 +44,21 @@ rxStack
...
-    stack = upConverterC'
-             |> asyncFifoC d4 ethClk ethRst ethEn hasClock hasReset hasEnable
-             |> preambleStripperC
-             |> fcsValidatorC
-             |> macDepacketizerC
-             -- TODO: filter on macAddressS or broadcoast
+    asyncFifoC' = asyncFifoC d4 ethClk ethRst ethEn hasClock hasReset hasEnable
+
+    isIpv4 = (== 0x0800) . _etherType
+    isForMyMac myMac h = let to = _macDst h in to == myMac || to == broadcastMac
```
Comment (rowanG077) [inline src/Clash/Cores/Ethernet/RxStack.hs]:
```Haskell
isForMyMac myMac (_macDst -> to) = to == myMac || to == broadcastMac
```

**Review by rowanG077 — CHANGES_REQUESTED** [review]
One change only really. And the splitting of the stack into mac and IP in examples @JLaumen mentioned.

File: src/Clash/Cores/Ethernet/Examples/RxStacks.hs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    upConverterC'
+    |> asyncFifoC'
+    |> preambleStripperC
+    |> fcsValidatorC
+    |> macDepacketizerC
+    |> filterMetaS (isForMyMac <$> macAddressS)
+  where
+    upConverterC' = exposeClockResetEnable upConverterC ethClk ethRst ethEn
+    asyncFifoC' = asyncFifoC d4 ethClk ethRst ethEn hasClock hasReset hasEnable
+    isForMyMac myMac (_macDst -> to) = to == myMac || to == broadcastMac
```
Comment (JLaumen) [inline src/Clash/Cores/Ethernet/Examples/RxStacks.hs:53]: 🥺👉👈

File: src/Clash/Cores/Ethernet/Examples/RxStacks.hs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  :: forall
+  (dataWidth :: Nat)
+  (dom :: Domain)
+  (domEth :: Domain).
+  ( HiddenClockResetEnable dom
+  , KnownDomain domEth
+  , HardwareCrc Crc32_ethernet 8 dataWidth
+  )
+  => 1 <= dataWidth
+  => KnownNat dataWidth
```
Comment (JLaumen) [inline src/Clash/Cores/Ethernet/Examples/RxStacks.hs]: I think this was already in the original RxStack like this, but formatting the constraints half as a tuple looks a bit weird. I'd prefer something like
```haskell
ipRxStack :: forall (dataWidth :: Nat) (dom :: Domain) (domEth :: Domain)
   . HiddenCLockResetEnable dom
  => KnownDomain domEth
  => HardwareCrc Crc32_ethernet 8 dataWidth
  => 1 <= dataWidth
  => KnownNat dataWidth
```
Or format it entirely as a tuple.

**Review by JLaumen — APPROVED** [review]

**Review by rowanG077 — APPROVED** [review]

**Pull request merged**